### DISCUSSION TITLE
How to make x-data reactive?

### DISCUSSION THREAD

**Discussion by catcombo** [alpinejs/alpine#4214] [1. Help] [answered]
I want to implement product page with "Add to cart" button. I store cart items in the Alpine store and want to get item for the product page based on the product ID and use it inside template. Here is a simple example https://jsfiddle.net/vo8h0rdj/

I fetch cart item from Alpine store in the `x-data` tag when page loads. I expected that if I update store (add or remove items) by clicking "Add"/"Remove" buttons current item will be updated automatically. But it doesn't. How to make it work?

**Comment by jy14898** [comment]
Not sure if this is the proper solution, but doing

```patch
- <div x-data="{ item: $store.cart.getItem(1) }">
+ <div x-data="{ get item() { return $store.cart.getItem(1); } }">
```

seems to make it work

**Reply by catcombo** [reply]
Yes, this will work. But the problem that it will call `getItem` on every access to the getter. I want to avoid it. In case if I have a big array, I would like to search for the item once, save it in the `item` variable and use it in template until array changes next time.

**Reply by ekwoka** [reply, marked as answer]
That isn't a major issue though.

But you can use watch/effect

```js
{
    item: null,
    init() {
      Alpine.effect(() => this.item = this.$store.cart.getItem(1))
    }
}
```

With more clear example, there might be a much better way to do this entirely, but having the getter go a few more times often won't be a huge issue. It's a cart, the array won't be 1000s of items.

**Reply by catcombo** [reply]
@ekwoka In general, I agree with you. But for every cart item I need to access price, quantity and total_price fields. It means I'll search for the item in the array 3 times per item. In case if I want to show a cart page, there will be 3 * number of items searches in the array. Of course, JavaScript engines in the browsers are very fast, devices fast as well. But I just can't write the code like this without caching/saving the data in the variable if I can do it somehow.

You suggestion works, thank you very much! I modified it a bit:
```javascript
<div x-data="{ item: undefined }" x-effect="item=$store.cart.getItem(1)">
```
https://jsfiddle.net/Lj57b4tw/

**Reply by ekwoka** [reply]
Why are you not just looping over the cart items directly?

How are you getting the cart item ids that you can't directly use the item itself?

This seems like a design issue.

**Reply by catcombo** [reply]
I use AlpineJS together with Django. I render everything I can in the backend. This is how product ID appears in the template. In my use case, AlpineJS is a handy tool that helps me avoid a lot of boilerplate code in pure JavaScript.

I agree with your opinion, it's better to use Alpine's `for` loop instead of rendering cart items in the backend. But in this case I'll have to change API to return not only the price, quantity and total_price fields, but also thumbnail URL, product name and other information I need to render product in the cart. Maybe I should do this.

My initial idea was to add some reactivity to my code, otherwise I could just create Angular or React app (with one small note, that I've never did it before :) ). Maybe I should reconsider how I use Alpine in my project.

**Reply by ekwoka** [reply]
Yeah, carts are very dynamic, and doing them wholly on the front end makes a lot of sense.

You can also in the store take the list of items and convert it to a dictionary or map so you can just lookup by id without looping.

**Reply by catcombo** [reply]
> You can also in the store take the list of items and convert it to a dictionary or map so you can just lookup by id without looping.

Yeah, I also thought about it. Anyway, thank you!
